src/net/socket.rs

```diff
@@ -0,0 +1,222 @@
+use crate::config::NetworkConfig;
+use crate::error::{NetworkError, NetworkErrorKind, NetworkResult};
+use crate::net::link_conditioner::LinkConditioner;
+use crate::net::{connection::ActiveConnections, events::SocketEvent};
+use crate::packet::Packet;
+use log::error;
+use mio::{Evented, Events, Poll, PollOpt, Ready, Token};
+use std::io;
+use std::mem;
+use std::net::{SocketAddr, ToSocketAddrs};
+use std::sync::{mpsc, Arc};
+
+const SOCKET: Token = Token(0);
+
+/// A reliable UDP socket implementation with configurable reliability and ordering guarantees.
+pub struct LaminarSocket {
+    socket: mio::net::UdpSocket,
+    config: Arc<NetworkConfig>,
+    connections: ActiveConnections,
+    recv_buffer: Vec<u8>,
```

Comment:
This was originally a `Vec` and I'm not entirely sure how I can have an array where the size is configurable at runtime.

Reply:
Why should the size be configurable at runtime? What is the use case around that?

Reply:
It should not be configurable. This is just the buffer the packets are written to. When we do a receive with UDP we get the size back, which we can use for indexing into this buffer. A vector is fine here since it is created once with the size from the config.

[Here](https://github.com/amethyst/laminar/blob/a3cd32bb7cddffe0e43633798e1dbd903243ec35/src/net/socket.rs#L187) the buffer is created once.

Reply:
I mean, I'm ok with locking it to the size of MTU (1500). However, if we do want it to be configurable, which it was originally, this must be a Vec

Reply:
If anything we don't need runtime configurable we need compile-time configurable buffer size.

Reply:
Don't hard code the MTU please. It'll kill performance when using this for simulations in clusters where jumbo frames work fine, for example.

Reply:
Definitely think leaving this as a `Vec` is the right move for now.

Reply:
I've opened an issue regarding this #141